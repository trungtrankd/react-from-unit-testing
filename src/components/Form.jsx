import React from "react";
import { Controller } from "react-hook-form";
import { withUseFormHook } from "../withUseHookForm";
import Checkbox from "./Checkbox";
import MultiSelect from "./MultiSelect";
import Options from "./Options";
import TextInput from "./TextInput";

// import Button from "./Button";

const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
class Form extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const languageList = ["English", "Spanish", "French", "German", "Japanese"];
    const { control, onSubmit, watch } = this.props;

    const languagesSelected = watch("languages");
    return (
      <form
        onSubmit={(e) => {
          e.preventDefault();
          onSubmit();
        }}
      >
        <Controller
          control={control}
          name="firstname"
          render={({ field, fieldState: { error } }) => (
            <>
              <TextInput
                labelFor="firstname"
                label="Firstname"
                id="firstname"
                {...field}
                className="form-control"
                type="text"
                data-testid="control-firstname"
              />
              {error && <p className="text-danger" data-testid="firstname-error">{error.message}</p>}
            </>
          )}
          rules={{
            required: {
              value: true,
              message: "Firstname is required",
            },
          }}
        />
        <Controller
          control={control}
          name="lastname"
          render={({ field, fieldState: { error } }) => (
            <>
              <TextInput
                labelFor="lastname"
                label="Lastname"
                {...field}
                className="form-control"
                type="text"
                id="lastname"
                data-testid="control-lastname"
              />
              {error && <p className="text-danger" data-testid="lastname-error">{error.message}</p>}
            </>
          )}
          rules={{
            required: {
              value: true,
              message: "Lastname is required",
            },
          }}
        />
        <Controller
          control={control}
          name="email"
          render={({ field, fieldState: { error } }) => (
            <>
              <TextInput
                labelFor="email"
                label="E-mail"
                className="form-control"
                type="email"
                id="email"
                {...field}
                data-testid="control-email"
              />
              {error && <p className="text-danger" data-testid="email-error">{error.message}</p>}
            </>
          )}
          rules={{
            required: {
              value: true,
              message: "E-mail is required",
            },
            pattern: {
              value: emailRegex,
              message: 'E-mail format invalid'
            }
          }}
        />
        <Controller
          name="languages"
          control={control}
          render={({ field }) => (
            <MultiSelect
              labelFor="languages"
              label="Languages"
              className="form-control mb-3"
              {...field}
              onChange={(e) => {
                const value = e.target.value;
                if (languagesSelected.some((x) => x === value)) {
                  field.onChange(languagesSelected.filter((x) => x !== value));
                } else {
                  field.onChange([...languagesSelected, value]);
                }
              }}
            >
              <Options options={languageList} />
            </MultiSelect>
          )}
        />
        <Controller
          control={control}
          name="subscribed"
          render={({ field }) => (
            <Checkbox
              labelFor="subscribed"
              label="Subscribe"
              checked={field.value}
              className="form-check mb-3 text-right"
              {...field}
            />
          )}
        />

        <button
          type="submit"
          data-testid="btn-submit"
          onClick={(e) => {
            e.preventDefault();
            onSubmit();
          }}
          className="btn btn-primary form-control"
        >
          Submit
        </button>
      </form>
    );
  }
}

export default withUseFormHook(Form);
