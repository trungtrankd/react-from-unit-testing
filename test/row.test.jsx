import { render } from "@testing-library/react";
import React from "react";
import Row from "./../src/components/Row";
import { describe, test } from "vitest";
import Column from './../src/components/Column';
import Form from './../src/components/Form';

describe("<Row />", () => {
  test("Should render correctly", () => {
    const component = render(<Row>
      <Column className="col-md-6 offset-md-3">
        <Form />
      </Column>
    </Row>);
    expect(component).toMatchSnapshot("./snapshot/row.output.html");
  });
});
