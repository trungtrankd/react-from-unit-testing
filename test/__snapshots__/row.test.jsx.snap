// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Row /> > Should render correctly > ./snapshot/row.output.html 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3"
        >
          <form>
            <div
              class="form-group"
            >
              <label
                for="firstname"
              >
                Firstname
              </label>
              <input
                class="form-control"
                data-testid="control-firstname"
                id="firstname"
                name="firstname"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="lastname"
              >
                Lastname
              </label>
              <input
                class="form-control"
                data-testid="control-lastname"
                id="lastname"
                name="lastname"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="email"
              >
                E-mail
              </label>
              <input
                class="form-control"
                data-testid="control-email"
                id="email"
                name="email"
                type="email"
                value=""
              />
            </div>
            <div>
              <label
                for="languages"
              >
                Languages
              </label>
              <select
                class="form-control mb-3"
                id="languages"
                multiple=""
                name="languages"
              >
                <option
                  value="English"
                >
                  English
                </option>
                <option
                  value="Spanish"
                >
                  Spanish
                </option>
                <option
                  value="French"
                >
                  French
                </option>
                <option
                  value="German"
                >
                  German
                </option>
                <option
                  value="Japanese"
                >
                  Japanese
                </option>
              </select>
            </div>
            <div
              class="form-check mb-3 text-right"
            >
              <input
                class="form-check-input"
                id="subscribed"
                name="subscribed"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="subscribed"
              >
                Subscribe
              </label>
            </div>
            <button
              class="btn btn-primary form-control"
              data-testid="btn-submit"
              type="submit"
            >
              Submit
            </button>
          </form>
          <div>
            <b>
              Result
            </b>
            <div
              class="firstname"
            >
              <span>
                Firstname:
              </span>
               
              <span
                class="data-firstname"
              />
               
            </div>
            <div
              class="lastname"
            >
              <span>
                Lastname:
              </span>
               
              <span
                class="data-lastname"
              />
               
            </div>
            <div
              class="email"
            >
              <span>
                E-mail:
              </span>
               
              <span
                class="data-email"
              />
               
            </div>
            <div
              class="languages"
            >
              <span>
                Languages:
              </span>
               
              <span
                class="data-languages"
              />
               
            </div>
            <div
              class="subscribed"
            >
              <span>
                subscribed:
              </span>
               
              <span
                class="data-subscribed"
              />
               
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
    >
      <div
        class="col-md-6 offset-md-3"
      >
        <form>
          <div
            class="form-group"
          >
            <label
              for="firstname"
            >
              Firstname
            </label>
            <input
              class="form-control"
              data-testid="control-firstname"
              id="firstname"
              name="firstname"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="lastname"
            >
              Lastname
            </label>
            <input
              class="form-control"
              data-testid="control-lastname"
              id="lastname"
              name="lastname"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="email"
            >
              E-mail
            </label>
            <input
              class="form-control"
              data-testid="control-email"
              id="email"
              name="email"
              type="email"
              value=""
            />
          </div>
          <div>
            <label
              for="languages"
            >
              Languages
            </label>
            <select
              class="form-control mb-3"
              id="languages"
              multiple=""
              name="languages"
            >
              <option
                value="English"
              >
                English
              </option>
              <option
                value="Spanish"
              >
                Spanish
              </option>
              <option
                value="French"
              >
                French
              </option>
              <option
                value="German"
              >
                German
              </option>
              <option
                value="Japanese"
              >
                Japanese
              </option>
            </select>
          </div>
          <div
            class="form-check mb-3 text-right"
          >
            <input
              class="form-check-input"
              id="subscribed"
              name="subscribed"
              type="checkbox"
            />
            <label
              class="form-check-label"
              for="subscribed"
            >
              Subscribe
            </label>
          </div>
          <button
            class="btn btn-primary form-control"
            data-testid="btn-submit"
            type="submit"
          >
            Submit
          </button>
        </form>
        <div>
          <b>
            Result
          </b>
          <div
            class="firstname"
          >
            <span>
              Firstname:
            </span>
             
            <span
              class="data-firstname"
            />
             
          </div>
          <div
            class="lastname"
          >
            <span>
              Lastname:
            </span>
             
            <span
              class="data-lastname"
            />
             
          </div>
          <div
            class="email"
          >
            <span>
              E-mail:
            </span>
             
            <span
              class="data-email"
            />
             
          </div>
          <div
            class="languages"
          >
            <span>
              Languages:
            </span>
             
            <span
              class="data-languages"
            />
             
          </div>
          <div
            class="subscribed"
          >
            <span>
              subscribed:
            </span>
             
            <span
              class="data-subscribed"
            />
             
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
